### ISSUE TITLE
A full run takes ~15 minutes for 20 prompts, against a 5-minute budget

### ISSUE THREAD

**Issue by ardsimm** [ardsimm/call_me_maybe#52]
# A full run takes ~15 minutes for 20 prompts, against a 5-minute budget

The subject requires processing all test prompts in under 5 minutes on standard hardware
(section V.5, "Reasonable speed"). A current run takes ~14 minutes for 20 prompts. A cProfile
run over the fixture set (11 prompts completed, 497.5 s) shows where the time goes, and it is
not where you would look first: **there is no slow code in `src/`**.

```
         3083920 function calls (2825629 primitive calls) in 497.511 seconds

   ncalls  tottime  cumtime  filename:lineno(function)
       12    0.000  495.605  call_me_maybe.py:128(__process_prompt)
      248    2.291  492.761  __generator_impl.py:85(__get_next_token)
      248    3.491  492.584  llm_sdk/__init__.py:93(get_logits_from_input_ids)
    48856  404.367  404.367  {built-in method torch._C._nn.linear}
     6944   44.383   44.383  {built-in method torch._C._nn.scaled_dot_product_attention}
       38    0.001  487.944  __generator_impl.py:118(__get_completion)
       11    0.001  370.926  __generator_impl.py:196(generate_parameters)
       11    0.004   96.838  __generator_impl.py:154(generate_name)
```

**492.6 s of 497.5 s (99.0%) is 248 calls to `get_logits_from_input_ids`**, at ~1.99 s each.
The constrainer, the trie, the states, the prompt building, the JSON adapter and every pydantic
validation together account for well under 1% of the wall clock. 75% of the run is parameter
generation, 20% is name selection.

`linear` is called 48856 times, which is 197 per forward pass: 28 layers x 7 projections, plus
`lm_head`. That is a complete forward pass over the entire prompt for every single generated
token, because the SDK exposes no KV cache.

So there are exactly two levers -- call the model fewer times, and make each call cheaper --
and the second one has a clean measured shape. Timing `get_logits_from_input_ids` against
prompt length on this machine (i7-12700, CPU only, no CUDA/MPS, AVX2 only, so `float32` on
12 torch threads):

| prompt tokens | median ms |
|---|---|
| 100 | 1107 |
| 150 | 1774 |
| 250 | 2059 |
| 375 | 3734 |
| 500 | 5247 |
| 600 | 5239 |

```
time_per_forward ~= 0.28 s + 0.0083 s x prompt_length
```

Linear. **Every token added to a prompt costs ~8 ms on every forward pass of every token
generated against that prompt.** Current prompt sizes, measured against the real fixtures with
12 functions: name selection 375 tokens, parameter generation 587 tokens (634 by the 5th
parameter of `fn_book_flight`).

## 1. One forward pass per completion is computed and discarded

`src/generate/__generator_impl.py:85`:

```python
token = constrainer.pick_token(
    self.model.get_logits_from_input_ids(result)
)
```

Python evaluates the argument before the call, so the logits are always computed. But
`ConstrainerImpl.pick_token` opens with:

```python
allowed_tokens = self.state.get_allowed_tokens()
if allowed_tokens is None:
    return None
```

When the state signals end-of-generation, `pick_token` returns immediately and the ~2 s forward
pass that produced those logits is thrown away. This happens exactly once per completion, at the
end of every one.

The arithmetic confirms it precisely rather than circumstantially: the run log contains 210
`Generated:` lines, and the profile shows 38 `__get_completion` calls and 248 forward passes.
210 + 38 = 248.

**38 of 248 forward passes -- 15.3% of total runtime -- produce logits that are never read.**

The state already knows it is finished before the model is consulted; it just is not asked until
after. The fix is to consult it first, either by having `__get_completion`/`__get_next_token`
check the constrainer's state before calling the model, or by giving `Constrainer` an accessor
for "is this state done" that the decode loop tests as its while-condition.

## 2. Forced tokens are put to the model anyway

`TrieState.get_allowed_tokens` returns `list(self.__current_node.keys())`. When that list holds
exactly one element the next token is fully determined: `max(allowed_tokens, key=logits.__getitem__)`
over a single-element list returns that element no matter what the logits say. The model is run
for ~2 s to answer a question with one possible answer.

For the 12 names in the current fixture:

| name | tokens | forced |
|---|---|---|
| fn_add_numbers | 3 | 2 |
| fn_int_len | 3 | 2 |
| fn_float_len | 3 | 2 |
| fn_greet | 3 | 2 |
| fn_reverse_string | 3 | 2 |
| fn_get_square_root | 4 | 3 |
| fn_substitute_string_with_regex | 6 | 5 |
| fn_send_message | 3 | 2 |
| fn_control_lights | 4 | 2 |
| fn_control_temperature | 3 | 1 |
| fn_call_taxi | 4 | 3 |
| fn_book_flight | 3 | 2 |
| **total** | **42** | **28 (67%)** |

**Two thirds of name-selection forward passes decide nothing.** The run log shows this directly:
for `fn_greet` the model is consulted for `fn`, `_g`, `reet` and `",` when only `_g` was ever a
real choice. The same applies to boolean parameters (`t` -> `rue` is forced) and to any
single-token transition in `IntState`, `FloatState` or `StringState`.

Worth being explicit, since it touches a hard constraint from the subject: skipping these does
**not** violate "the function to call should be chosen using the LLM, not with heuristics". The
LLM still makes every branching decision in the trie. What is skipped is the deterministic tail
of a path the LLM has already chosen -- the outcome is byte-for-byte identical, only faster.

## 3. The parameter prompt is roughly twice the size it needs to be

`data/templates/function_parameters/function_parameters_main_template.txt` is 587 tokens:

| part | tokens |
|---|---|
| system header (the Rules block) | 137 |
| few-shot example 0 (`fn_add_number`) | 78 |
| few-shot example 1 (`fn_rate_product`) | 98 |
| few-shot example 2 (`fn_greet`) | 65 |
| few-shot example 3 (`fn_rate_product`) | 98 |
| the actual request | ~110 |

Two of those are free to remove:

**Few-shot examples 1 and 3 are byte-identical.** The `fn_rate_product` /
"Stopped working after a week" block appears twice, verbatim. 98 tokens x ~8 ms, on every forward
pass of every parameter of every prompt, teaching the model nothing it did not already learn from
the first copy.

**`Prompting.__context_prompt` is redundant.** `src/prompting/prompting.py:8` prepends a 32-token
`<|im_start|>system ... <|im_end|>` block to prompts whose templates already open with their own
`<|im_start|>system` block. Both the name and parameter prompts therefore carry two system
messages with different instructions -- a correctness smell as much as a cost, and it is paid on
every forward pass of the entire run.

Those two cuts alone take the parameter prompt from 587 to 457 tokens (-22%) with no judgement
call involved. Going further -- 4 few-shot examples down to 2, keeping one string-with-escaped-quote
and one numeric -- reaches ~280 tokens (-52%), i.e. 5.15 s -> 2.35 s per forward pass, a 2.2x
speedup on the 75% of runtime spent generating parameters. That last cut is the only change in
this issue that can move accuracy, so it should be measured separately and last.

The name prompt's 375 tokens are mostly the 12 function descriptions, which are load-bearing.
Leave them alone.

## What is deliberately not worth optimising

Recorded so no time is spent here later:

- `logits.index(max(logits))` in `ConstrainerImpl`'s unconstrained branch is two Python passes
  over 151936 floats: measured 9.8 ms, ~1.5 s across the whole run (0.3%). `numpy.argmax` would
  fix it and numpy is permitted, but it buys nothing meaningful.
- `out.logits[0, -1].tolist()` and the `[float(x) for x in logits]` comprehension inside the SDK:
  3.5 s total (0.7%).
- The `print("Generated:", ...)` in `ConstrainerImpl.pick_token`: one tokenizer decode per token,
  negligible.
- Trie construction, pydantic validation, template loading, JSON I/O: all below 0.1 s.

## The structural ceiling, and what to say about it in the README

Even after everything above, the design re-prefills the whole prompt for every generated token.
A 280-token prompt producing 12 tokens processes 3360 token-positions where a KV cache would
process 292 -- ~12x more work than necessary. Two thirds of the `lm_head` work is wasted as well,
since transformers computes logits for every position (`logits_to_keep` defaults to 0) when only
the last row is ever read.

Neither is reachable through the permitted surface: `get_logits_from_input_ids` takes a `list[int]`
and returns logits, with no way to pass or receive `past_key_values`, and `self._model` is a
private attribute the subject forbids touching. This belongs in the README's "Performance analysis"
and "Challenges faced" sections -- the per-token full prefill is a constraint imposed by the
provided SDK, and the optimisation work consisted of minimising both the number of prefills and
their length.

## Projected result

Starting from 45.2 s per prompt:

| change | effect | s/prompt |
|---|---|---|
| baseline | | 45.2 |
| item 1 (skip the discarded final forward) | -15% | 38.3 |
| item 2 (skip forced tokens) | name gen 97 s -> 30 s | 32.2 |
| item 3, dup few-shot + context block | param forwards -22% | 26.8 |
| item 3, 4 shots -> 2 | param forwards -25% more | ~20 |

~20 s per prompt, so 20 prompts in ~6.5 minutes: a 2.3x improvement, still slightly over budget.
Closing the last gap means either a more aggressive trim of the parameter prompt (the 137-token
Rules block is the next target) or running prompts in parallel processes -- the subject's bonus
list does name "performance optimizations (caching, batching)". Caveats on the parallel option:
each process loads its own ~2.4 GB of `float32` weights against ~7 GB free here, so 2-3 workers
maximum; torch already uses 12 threads, so workers need `torch.set_num_threads` lowered to avoid
oversubscription; and output ordering must be restored before writing the JSON array.

## Checklist

- [x] `__get_next_token` (`src/generate/__generator_impl.py:85`) passes
      `self.model.get_logits_from_input_ids(result)` as an argument to `pick_token`, so the logits
      are always computed even though `pick_token` may return `None` without reading them. Consult
      the state for completion *before* calling the model. Confirmed to be exactly one wasted
      forward pass per completion (210 generated tokens + 38 completions = 248 forward passes).
      Worth 15.3% of runtime.
- [x] When `state.get_allowed_tokens()` holds exactly one token, return it without calling the
      model -- the argmax over a one-element list cannot depend on the logits. 28 of 42 name
      tokens (67%) in the current fixture are forced this way, plus `rue` after `t` for booleans.
      Verify the generated names are byte-for-byte identical before and after; the LLM must still
      make every branching choice.
- [ ] Delete the duplicated few-shot example in
      `data/templates/function_parameters/function_parameters_main_template.txt` -- the
      `fn_rate_product` / "Stopped working after a week" block appears twice, verbatim. 98 tokens
      off every parameter forward pass.
- [ ] Remove `Prompting.__context_prompt` (`src/prompting/prompting.py:8`), or fold it into the
      templates. It prepends a second `<|im_start|>system` block to templates that already open
      one, so the model receives two competing system messages at a cost of 32 tokens on every
      forward pass of the run.
- [ ] Re-profile after the four items above and record the new s/prompt in `claude/metrics/`
      before touching anything else, so the accuracy-neutral gains are measured in isolation.
- [ ] Only then: trim the parameter template's few-shot examples from 4 to 2 (keep one
      string-with-escaped-quote, one numeric) and consider compressing the 137-token Rules block.
      Re-run the `claude/test_cases/` scenarios and record accuracy alongside speed -- this is the
      one change here that can regress correctness.
- [ ] Document the per-token full prefill and the unreachable KV cache in the README's
      "Performance analysis" and "Challenges faced" sections, with the measured
      `0.28 s + 0.0083 s x prompt_length` relationship as the justification for why prompt length
      was the thing optimised.

*Issue found and written by claude*
